maestro-orchestra/src/main/java/maestro/orchestra/Orchestra.kt

```diff
@@ -113,6 +113,45 @@ class Orchestra(
 
     private val rawCommandToMetadata = mutableMapOf<MaestroCommand, CommandMetadata>()
 
+    private var isStopping = false
+    private var isPaused = false
+    private val pauseLock = Object()
+
+    fun stopFlow() {
+        isStopping = true
+
+        synchronized(pauseLock) {
+            pauseLock.notifyAll()
+        }
+    }
+
+    fun pauseFlow() {
+        synchronized(pauseLock) {
+            isPaused = true
+            pauseLock.notifyAll() 
+        }
+    }
+
+    fun resumeFlow() {
+        synchronized(pauseLock) {
+            isPaused = false
+            pauseLock.notifyAll()
+        }
+    }
+
+    private fun checkPauseState() {
+        synchronized(pauseLock) {
```

Comment:
If we plan to merge the coroutine implementation we can do this via coroutine:

Something like this:

```kotlin
private suspend fun waitWhilePaused() {
        // suspend this block until the variable isPause is true
       while (isPaused) {
            delay(pauseCheckDelay)
      }
}

    // Example of a coroutine with pause functionality
suspend fun executeCommands() {
      try {
          // command execution
          waitWhilePaused()
       } catch (e: CancellationException) {
          println("Coroutine was cancelled")
          throw e
       }
}
```